### PR TITLE
[RLlib] Fix env_check for parametric actions (with action mask)

### LINKED ISSUE #23925: [RLlib] check_gym_environments not working with parametric actions
### What happened + What you expected to happen

The new `check_gym_environments` script in 2.3.1 is not working with a recommended way to deal with parametric actions (https://docs.ray.io/en/latest/rllib/rllib-models.html#variable-length-parametric-action-spaces).

The action space of a parametric environment is still discrete, so the checking script expects a step to succeed with any random action from it, which is not the case since some of them are not available.

For now, the only solution is to disable this check even for the environments for which, ideally, it should pass without errors.

### Versions / Dependencies

2.3.1

### Reproduction script

```python
from ray.rllib.examples.env.parametric_actions_cartpole import (
    ParametricActionsCartPole,
)
from ray.rllib.utils.pre_checks.env import check_gym_environments

parametric_env = ParametricActionsCartPole(1000)
check_gym_environments(parametric_env)
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by inpefess**
## Why are these changes needed?

Ray RLlib has a great [example](https://github.com/ray-project/ray/blob/master/rllib/examples/parametric_actions_cartpole.py) of using a parametric actions environment, but now it works only with `self._skip_env_checking = True`. Gymnasium action spaces have a `mask` argument to their `sample` method. We apply this feature to fix the `env_check` behaviour in the parametric actions environments case.

## Related issue number

Closes  #23925

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

## Discussion

1. I added a test to `rllib/utils/tests/test_env_check.py` because it seems to belong there, but seven tests from this module fail in the `master` branch
2. The assumption that having a dictionary with the "action_mask" key is a standard parametric actions implementation seems too strong (one can use other names or add the mask to `info`, for example). If it's a standard or recommended way to do that in Ray RLlib, then one should also mention it in the documentation (now it's not mentioned at all, although the previous version of the documentation page was a bit more verbose). It seems to be a "standard" way, for example, in AlphaZero it's the same and in a couple of other examples.
3. Apart from `ParametricActionsCartPole` using the `action_mask` key, there is a `ParametricActionsCartPoleNoEmbeddings` having a key with the same meaning but called `valid_avail_actions_mask`. I've changed it to `action_mask`

### CODE REVIEW

**Comment by inpefess** [comment]
@avnishn Could you review this PR, please? Failing checks don't seem to be related to the changes proposed.

**Review by avnishn — APPROVED** [review]

**Comment by avnishn** [comment]
> The assumption that having a dictionary with the "action_mask" key is a standard parametric actions implementation seems too strong (one can use other names or add the mask to info, for example). If it's a standard or recommended way to do that in Ray RLlib, then one should also mention it in the documentation (now it's not mentioned at all, although the previous version of the documentation page was a bit more verbose). It seems to be a "standard" way, for example, in AlphaZero it's the same and in a couple of other examples.

I think that is very much the case. We have to think about if we want to make an opinionated design decision on how to support action masking in RLlib. 

On an initial thought, I think that we'd rather not make an opinionated design decision here. I'll talk with @sven1977 who probably has more opinions on this and get back to you.

File: rllib/utils/tests/test_check_env.py [hunk 1]
```diff
@@ -104,6 +106,10 @@ def test_step(self):
         with pytest.raises(ValueError, match=error):
             check_env(env)
 
+    def test_parametric_actions(self):
```
Comment (sven1977) [inline rllib/utils/tests/test_check_env.py:109]: Very nice!

File: rllib/utils/pre_checks/env.py [hunk 2]
```diff
@@ -212,6 +213,10 @@ def check_gym_environments(env: Union[gym.Env, "old_gym.Env"]) -> None:
                     space_type,
                 )
             )
+    # sample a valid action in case of parametric actions
+    if isinstance(reset_obs, dict):
```
Comment (sven1977) [inline rllib/utils/pre_checks/env.py:217]: This is a really cool feature of gymnasium, actually :) which I didn't know about.

**Review by sven1977 — COMMENTED** [review]
Awesome PR. Thanks for filing this @inpefess . I would like to suggest one enhancement to make the "action_mask" key not hard-coded.
Can we add an additional config setting in the`rllib/algorithms/algorithm_config.py::AlgorithmConfig::environment()` method to be able to customize the actual value of this action mask key in the observation_space?

something along the lines of:
```
config.environment("my_env", env_config=..., action_mask_key="action_mask")
```

Set the default value of `self.action_mask_key = "action_mask"` in the AlgorithmConfig c'tor.

Then use that value instead of the hard-coded one in the pre-check.
You might have to change the signature of `check_env` to pass along the AlgorithmConfig (from within the RolloutWorker) so that it has access to that configuration.

File: rllib/algorithms/algorithm_config.py [hunk 3]
```diff
@@ -1377,6 +1376,9 @@ def environment(
                 (gym.wrappers.EnvCompatibility). If False, RLlib will produce a
                 descriptive error on which steps to perform to upgrade to gymnasium
                 (or to switch this flag to True).
+             action_mask_key: If observation is a dictionary, expect the value by
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:1379]: Awesome, thanks for adding this so quickly! I think it's ready to merge now. Just waiting for tests to finish ..

**Review by sven1977 — APPROVED** [review]
Looks great now. Thanks again @inpefess !

**Comment by afennelly-mitre** [comment]
@inpefess @sven1977 @avnishn I was browsing through this PR from earlier this year, and wanted to verify if my assumption is correct about the changes in this PR: 

the env_check for parametric actions (with action mask) will **only** work if the underlying environment is a `gym.Env`, and will not work if the environment is say, a `VectorEnv` or `MultiAgentEnv` since the logic is only added to the [check_gym_environments()](https://github.com/ray-project/ray/blob/64e53731df4fad877da6e6ceef9ac432e0e85ad0/rllib/utils/pre_checks/env.py#L109) method?

**Pull request merged**